R/compat-obj-type.R

```diff
@@ -59,44 +66,62 @@ obj_type_friendly <- function(x, value = TRUE, length = FALSE) {
 
   n_dim <- length(dim(x))
 
-  if (value && !n_dim) {
-    if (is_na(x)) {
-      return(switch(
-        typeof(x),
-        logical = "`NA`",
-        integer = "an integer `NA`",
-        double =
-          if (is.nan(x)) {
-            "`NaN`"
-          } else {
-            "a numeric `NA`"
-          },
-        complex = "a complex `NA`",
-        character = "a character `NA`",
-        .rlang_stop_unexpected_typeof(x)
-      ))
-    }
-    if (length(x) == 1 && !is_list(x)) {
-      return(switch(
-        typeof(x),
-        logical = if (x) "`TRUE`" else "`FALSE`",
-        integer = "an integer",
-        double =
-          if (is.infinite(x)) {
-            if (x > 0) {
-              "`Inf`"
+  if (!n_dim) {
+    if (!is_list(x) && length(x) == 1) {
+      if (is_na(x)) {
+        return(switch(
+          typeof(x),
+          logical = "`NA`",
+          integer = "an integer `NA`",
+          double =
+            if (is.nan(x)) {
+              "`NaN`"
             } else {
-              "`-Inf`"
-            }
+              "a numeric `NA`"
+            },
+          complex = "a complex `NA`",
+          character = "a character `NA`",
+          .rlang_stop_unexpected_typeof(x)
+        ))
+      }
+
+      show_infinites <- function(x) {
+        if (x > 0) {
+          "`Inf`"
+        } else {
+          "`-Inf`"
+        }
+      }
+
+      if (value) {
+        if (is.numeric(x) || is.complex(x)) {
+          if (is.infinite(x)) {
+            return(show_infinites(x))
```

Comment:
You might need a special path for complex because you can't compare them

``` r
complex(real = 1, imaginary = Inf) > 0
#> Error in complex(real = 1, imaginary = Inf) > 0: invalid comparison with complex values
```